File: test/ctest/case/chain_api.cpp
#include "ctest.h"
#include "gtest/gtest.h"

class ChainAPI : public testing::Test{
protected:
	void test();
	bumo::CTest ctest_;
};

TEST_F(ChainAPI, test){ test(); }

void ChainAPI::test(){
	//issue asset
	bumo::CTest::Param param;
	param.balance = "1000000000000000000";
	param.fee_limit = "20000000000";
	param.gas_price = "1000";
	param.input = "init";

	param.asset_key = "bucoin";
	param.asset_issuer = ctest_.GetConfig().address_;
	EXPECT_EQ(ctest_.Issueasset(param), 0);

	param.balance = "10000000000";
	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);

	//getAccountMetadata
	param.input = "getBalance";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("getBalance-0"), "0");
	ASSERT_EQ(ctest_.Load("getBalance-1"), "20000000000");

	//load
	param.input = "load";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("init-main-0"), "init");
	EXPECT_EQ(ctest_.Load("init-main-1"), "false");

	//del
	param.input = "del";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("init-main-0"), "");
	EXPECT_EQ(ctest_.Load("init-main-1"), "");

	//getBlockHash
	param.input = "getBlockHash";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_NE(ctest_.Load("getBlockHash"), "");

	//tlog
	param.input = "tlog";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("tlog"), "true");
	
	//getValidators
	param.input = "getValidators";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_NE(ctest_.Load("getValidators"), "");

	//getAccountMetadata
	param.input = "getAccountMetadata";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("getAccountMetadata-0"), "false");
	ASSERT_EQ(ctest_.Load("getAccountMetadata-1"), "init");

	//getAccountAsset
	param.input = "payasset";
	param.asset_key = "bucoin";
	param.asset_issuer = ctest_.GetConfig().address_;
	EXPECT_EQ(ctest_.Payasset(param), 0);
	param.input = "getAccountAsset";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("getAccountAsset-0"), "false");
	ASSERT_EQ(ctest_.Load("getAccountAsset-1"), "false");
	ASSERT_EQ(ctest_.Load("getAccountAsset-2"), param.balance);

	//configFee
	param.input = "configFee";
	EXPECT_EQ(ctest_.Paycoin(param), 151);

	//setValidators
	param.input = "setValidators";
	EXPECT_EQ(ctest_.Paycoin(param), 151);

	//payCoin
	param.input = "payCoin";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("getBalance-3"), param.balance);

	//payAsset
	param.input = "payAsset";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	ASSERT_EQ(ctest_.Load("payAsset"), "20000000");

	//Write in the inter_call file
	//delegateCall£¬delegateQuery
	//contractCall,contractQuery
}

File: test/ctest/case/chain_api.js
'use strict';

function init(input_str) {
    Chain.store('init', input_str);
    Utils.assert(Chain.msg.asset === undefined);
}

function main(input_str) {

    if (input_str === 'getBalance') {
        let balance_1 = Chain.getBalance('1231');
        Chain.store('getBalance-0', balance_1);

        let balance_2 = Chain.getBalance(Chain.thisAddress);
        Chain.store('getBalance-1', balance_2);
    }

    if (input_str === 'load') {
        let value = Chain.load('init');
        Chain.store('init-main-0', value);

        value = Chain.load('init-null');
        Chain.store('init-main-1', value.toString());
    }

    if (input_str === 'del') {
        Chain.del('init-main-0');
        Chain.del('init-main-1');
    }

    if (input_str === 'getBlockHash') {
        let hash = Chain.getBlockHash(0);
        Chain.store('getBlockHash', hash);
    }

    if (input_str === 'tlog') {
        let tlogResult = Chain.tlog('tlog', 'bbbbbbbbbbbbbb');
        Chain.store('tlog', tlogResult.toString());
    }

    if (input_str === 'getValidators') {
        let validators = Chain.getValidators();
        Chain.store('getValidators', JSON.stringify(validators));
    }

    if (input_str === 'getAccountMetadata') {
        let meta_1 = Chain.getAccountMetadata(Chain.thisAddress, 'hello');
        Chain.store('getAccountMetadata-0', meta_1.toString());

        let meta_2 = Chain.getAccountMetadata(Chain.thisAddress, 'init');
        Chain.store('getAccountMetadata-1', meta_2.toString());
    }

    if (input_str === 'payasset') {
        Chain.store('asset-key', JSON.stringify(Chain.msg.asset));
    }

    if (input_str === 'getAccountAsset') {
        let asset_0 = Chain.getAccountAsset(Chain.thisAddress, 'hello');
        Chain.store('getAccountAsset-0', asset_0.toString());

        let asset_json = { 'issuer': 'buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY', 'code': 'bu' };
        let asset_1 = Chain.getAccountAsset(Chain.thisAddress, asset_json);
        Chain.store('getAccountAsset-1', asset_1.toString());

        let asset_json_2 = JSON.parse(Chain.load('asset-key'));
        let asset_2 = Chain.getAccountAsset(Chain.thisAddress, asset_json_2.key);
        Chain.store('getAccountAsset-2', asset_2);
    }

    if (input_str === 'configFee') {
        Chain.configFee('hello');
    }

    if (input_str === 'setValidators') {
        Chain.setValidators('hello');
    }

    if (input_str === 'payCoin') {
        let balance_3 = Chain.getBalance('buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY');
        Chain.payCoin('buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY', Chain.msg.coinAmount);
        let balance_4 = Chain.getBalance('buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY');
        Chain.store('getBalance-3', int64Sub(balance_4, balance_3));
    }

    if (input_str === 'issueAsset' || input_str === 'payAsset') {
        Chain.issueAsset('bu-test', '30000000');
        Chain.payAsset('buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY', Chain.thisAddress, 'bu-test', '20000000');

        let asset_json_3 = { 'issuer': '', 'code': 'bu-test' };
        asset_json_3.issuer = Chain.thisAddress;
        let asset_3 = Chain.getAccountAsset('buQsZNDpqHJZ4g5hz47CqVMk5154w1bHKsHY', asset_json_3);
        Chain.store('payAsset', asset_3);
    }

    return '';
}

function query(input_str) {
    return '';
}
